src/torchio/datasets/slicer.py

```diff
@@ -36,7 +36,7 @@
 }
 
 
-class Slicer(_RawSubjectCopySubject):
+class Slicer(Subject):
```

Comment:
nit: this changes behavior where it's not necessary.

before a copy of this class was of type `Subject` now i's only a instance of a subclass.

It's fine by me in general, I just remember these mechanics to be required like that in the past. Thoughts here @fepegar ?

Reply:
@justusschock Thanks for taking a look at the PR!

 The `_RawSubjectCopySubject` inherits from `Subject` and implements **only** the following:
```python
class _RawSubjectCopySubject(Subject):
    def __copy__(self):
        return _subject_copy_helper(self, Subject)
```

I removed the entire class as 
1. It already implemented exactly what `Subject` did, i.e. a custom `__copy__`, with the exact same logic using `_subject_copy_helper`
2. As discussed in the entire issue, the custom `__copy__` with a mix of deep and shallow copies resulted in a memory leak

Based on the logic of the custom `__copy__` method, it appears we want to protect the `Image` and ultimately the mutable tensors in each of them as we are manually creating a deep copy just for `Image` but keeping all other attributes as a shallow copy regardless of whether they are mutable or not. I figured that mixing both in the shallow copy dunder/magic method most likely resulted in the memory leak, which has been shown in #1222 . To create a deep copy we can just rely on the regular `copy.deepcopy` without the requirement of a custom logic, particularly both `Subject` and `Image` are of type Dict and just copy each of the key-value pairs. 

To check for it, I implemented two new tests for both shallow and deep copies. You will see that both new tests pass as expected. In addition, everywhere we are creating a shallow copy of the subject I changed it to a deep copy to really protect it.

If we should have `copy=True` as a default argument to `Transform` might be a different discussion as we never change a tensor (an `Image`) on disk so we never really change it permanently anyways, which might make the copy itself redundant.

Reply:
> If we should have `copy=True` as a default argument to `Transform` might be a different discussion as we never change a tensor (an `Image`) on disk so we never really change it permanently anyways, which might make the copy itself redundant.

I do not fully understand, this argument: Of course we do not want to change the file on disk. The issue is only about Subject variable in a python script. I advocate for keeping `copy=True` as default value, because most of the time when we wrote 
`subject_transformed = transform(subject)`
on expect to have 2 different instance of `Subject` (unless we explicitly ask for `copy=False`)
... no ?

Reply:
Well I figured that the copy of the `Subject` is in place to protect the initial state of the `Subject` prior to any modification. I am specifically refering to any `Transform`, that may alter the content of the `Subject` and more precisely the `Image`, which usually is loaded from disk (multiple nii.gz or similar medical images). However, every modification is only temporary as we **never** modify the true origin of any `Subject`, i.e. the original image on disk is never replaced in the code unless the user actively calls `Image.save`. The question/thought I had was why protect something that we never permanently change? We always reload the ground-truth from disk whenever we access any `Image`/`Subject`. So why have `copy=True` as a default, that may or may not incur a memory leak as shown in #1222 without any real purpose?

Reply:
Not sure ...
 The real purpose of copy=True is not to protect the file on disk but only to protect the loaded subject so that you can write 
```
#with transfrom being any torchio transform
subject_transformed = tranform(subject)
```
with copy=True default (in the transform) then subject_transformed and subject have now two different contents

This is the default behaviour we want to keep, and therefore I think it is important to keep copy=True by default ...

Reply:
Maybe I have an issue in my thinking. Could you explain to me why it is important to do a copy by default of a `Subject`, if we never save the initial subject and never return the initial subject, which means it will be just deleted by the garbage collector eventually? We are just returning the modified/transformed `Subject`, forget about the initial one and keep progressing in the code. Why not do the tensor operations in-place? Why "protect" something we do not care about anymore?

Reply:
ok, this is probably just a matter of the exact use case. I can imagine cases where we need to keep the initial state of subject's images after the transform in order to **access it latter** in the python script (without reloading it from disk). For instance if one wish to compare the image before and after a motion artefact. 
So my point was just to say that ` copy=True ` is not useless .

this being said, I am not sure if this should the it be the default ... 

Anyway we do agree, that it is important to make the transform compliant to the user choice (copy=True or False) so indeed #1223 need to be fixed .